packages/SwingSet/src/kernel/kernel.js

```diff
@@ -683,26 +681,25 @@ export default function buildKernel(
       }
       let didAbort = false;
       if (terminationTrigger) {
-        const {
-          vatID,
-          shouldAbortCrank,
-          shouldReject,
-          info,
-        } = terminationTrigger;
-        if (shouldAbortCrank) {
+        // the vat is doomed, either voluntarily or from meter/syscall fault
+        const { vatID, shouldReject, info } = terminationTrigger;
+        if (terminationTrigger.shouldAbortCrank) {
+          // errors unwind any changes the vat made
           abortCrank();
           didAbort = true;
         }
+        // state changes reflecting the termination must survive, so these
+        // happen after a possible abortCrank()
         terminateVat(vatID, shouldReject, info);
         kernelSlog.terminateVat(vatID, shouldReject, info);
         kdebug(`vat terminated: ${JSON.stringify(info)}`);
       }
       if (!didAbort) {
-        kernelKeeper.processRefcounts();
-        kernelKeeper.saveStats();
         // eslint-disable-next-line no-use-before-define
         await vatWarehouse.maybeSaveSnapshot();
       }
+      kernelKeeper.processRefcounts();
```

Comment:
It's unclear to me what effect `processRefCounts` will have in the even of an aborted vat, but my intuition is that it's noise at best -- when a vat aborts, you need to clean up everything it left behind, and it's not clear to me that `processRefCounts` will do the right thing in such a case.  Possibly you've considered this, so calibrate my comment accordingly.

Reply:
The commit comment for that change has a description of the bug I think we had.. take a look at that and see if it makes sense to you. I should really write a test to exercise that, but I'm in a rush. I'll make a note to add one.

Reply:
Ah, the PR was so small I didn't notice that it was two commits with separate comments.  Yeah, the commit comment makes sense to me.  The key thing is that the DB changes for the abort are discarded before the DB changes for the termination cleanup begin.

BTW, as for stats, I share your concern that they may be a little messed up by this, but the stats themselves should be regarded as bit flakey and non-authoritative (and subject to arbitrary changes at arbitrary times, in terms of both what is collected and when) and shouldn't be part of the consensus state in any case.